### ISSUE TITLE
standalone/io/process_start_exception_test : sometimes wrong error code on android

### ISSUE THREAD

**Issue by jensjoha** [dart-lang/sdk#28299]
```
FAILED: precompiler-dart_precompiled release_arm standalone/io/process_start_exception_test
Expected: Pass 
Actual: RuntimeError
CommandOutput[precompiler]:
CommandOutput[adb_precompilation]:

stdout:
Executing adb -s 0604df07003b8e0a shell rm -Rf /data/local/tmp/precompilation-testing/test ; echo AdbShellExitCode:  $?
Stdout:
AdbShellExitCode: 0
ExitCode: 0
Time: 0:00:00.057219

Executing adb -s 0604df07003b8e0a shell mkdir -p /data/local/tmp/precompilation-testing/test ; echo AdbShellExitCode:  $?
Stdout:
AdbShellExitCode: 0
ExitCode: 0
Time: 0:00:00.052727

Executing adb -s 0604df07003b8e0a push out/ReleaseAndroidARM/dart_precompiled_runtime /data/local/tmp/precompilation-testing/dart_precompiled_runtime
Stderr:
3447 KB/s (59867140 bytes in 16.958s)
ExitCode: 0
Time: 0:00:17.040601

Executing adb -s 0604df07003b8e0a push out/ReleaseAndroidARM/process_test /data/local/tmp/precompilation-testing/process_test
Stderr:
2090 KB/s (78380 bytes in 0.036s)
ExitCode: 0
Time: 0:00:00.123228

Executing adb -s 0604df07003b8e0a shell chmod 777 /data/local/tmp/precompilation-testing/dart_precompiled_runtime /data/local/tmp/precompilation-testing/process_test ; echo AdbShellExitCode:  $?
Stdout:
AdbShellExitCode: 0
ExitCode: 0
Time: 0:00:00.109342

Executing adb -s 0604df07003b8e0a push /b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/out/ReleaseAndroidARM/generated_compilations/precompiler/tests_standalone_io_process_start_exception_test/out.aotsnapshot /data/local/tmp/precompilation-testing/test/out.aotsnapshot
Stderr:
3417 KB/s (3652784 bytes in 1.043s)
ExitCode: 0
Time: 0:00:01.128810

Executing adb -s 0604df07003b8e0a shell /data/local/tmp/precompilation-testing/dart_precompiled_runtime --ignore-unrecognized-flags --packages=/b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/.packages /data/local/tmp/precompilation-testing/test/out.aotsnapshot ; echo AdbShellExitCode:  $?
Stdout:
Unhandled exception:
Expect.equals(expected: <2>, actual: <13>, 'ProcessException: Permission denied
  Command: __path_to_something_that_should_not_exist__ ') fails.
#0      Expect.equals (package:expect/expect.dart)
#1      testStartError.<anonymous closure> (file:///b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/tests/standalone/io/process_start_exception_test.dart)
#2      _RootZone.runUnary (dart:async/zone.dart)
#3      _FutureListener.handleError (dart:async/future_impl.dart)
#4      _Future._propagateToListeners.handleError (dart:async/future_impl.dart)
#5      _Future._propagateToListeners (dart:async/future_impl.dart)
#6      _Future._completeError (dart:async/future_impl.dart)
#7      _Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart)
#8      _microtaskLoop (dart:async/schedule_microtask.dart)
#9      _startMicrotaskLoop (dart:async/schedule_microtask.dart)
#10     _Timer._runTimers (dart:isolate-patch/timer_impl.dart)
#11     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart)
#12     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart)
AdbShellExitCode: 255
ExitCode: 255
Time: 0:00:00.153389



Command[precompiler]: DART_CONFIGURATION=ReleaseAndroidARM out/ReleaseAndroidARM/dart_bootstrap --snapshot-kind=app-aot --snapshot=/b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/out/ReleaseAndroidARM/generated_compilations/precompiler/tests_standalone_io_process_start_exception_test/out.aotsnapshot --use-blobs --no-sim-use-hardfp --ignore-unrecognized-flags --packages=/b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/.packages /b/build/slave/vm-precomp-android-release-2-3-be/build/sdk/tests/standalone/io/process_start_exception_test.dart
Took 0:00:02.804601

Command[adb_precompilation]: Steps to push precompiled runner and precompiled code to an attached device. Uses (and requires) adb.
Took 0:00:18.774251

Short reproduction command (experimental):
    python tools/test.py -mrelease -cprecompiler -rdart_precompiled -aarm -sandroid --use-blobs -t480 standalone/io/process_start_exception_test
```

with the test basically looking like this:
```dart
testRunError() {
  Future<ProcessResult> processFuture =
      Process.run("__path_to_something_that_should_not_exist__",
                  const []);

  processFuture.then((result) => Expect.fail("exit handler called"))
  .catchError((error) {
    Expect.isTrue(error is ProcessException);
    Expect.equals(2, error.errorCode, error.toString());
  });
}
```
we get error code 13 instead of 2.

Happened in both
https://build.chromium.org/p/client.dart/builders/vm-precomp-android-release-2-3-be/builds/844/steps/test%20vm/logs/stdio
and
https://build.chromium.org/p/client.dart/builders/vm-precomp-android-release-2-3-be/builds/861/steps/test%20vm/logs/stdio

**Comment by jensjoha** [comment]
https://codereview.chromium.org/2623493002/

**Comment by mraleph** [comment]
This test flaked again. I am gonna add some printing to try and diagnose this next time it fails.

**Comment by mraleph** [comment]
I figured this one out - the test is wrong, it does not account for a possibility that `PATH` environment variable contains directory that is inaccessible by a normal user.

```
$ adb shell
shell@bullhead:/ $ echo $PATH
/sbin:/vendor/bin:/system/sbin:/system/bin:/system/xbin
shell@bullhead:/ $ ls -al /sbin
opendir failed, Permission denied
```

I looked at `execvp` implementation and it returns `EACCES` if it tried to search for the binary in directories listed within `PATH` *and* it did not find any suitable binary *and* at least one of the directories were not accessible - which fits the bill. 

I will fix the test to account for this. It's not entirely clear why this test is flaky though - probably some phones have something like `/sbin` in the `PATH` and some don't.
